src/Microsoft.OData.Core/ODataBinaryStreamWriter.cs

```diff
@@ -131,37 +134,40 @@ public override long Position
         /// <param name="count">The number of bytes to write.</param>
         public override void Write(byte[] bytes, int offset, int count)
         {
-            byte[] prefixByteString = emptyByteArray;
-            int trailingByteLength = trailingBytes.Length;
-            int numberOfBytesToPrefix = trailingByteLength > 0 ? 3 - trailingByteLength : 0;
+            Debug.Assert(this.wrappedBuffer != null, "this.wrappedBuffer != null");
 
             // if we have less than 3 bytes, store the bytes and continue
-            if (count + trailingByteLength < 3)
+            if (count + trailingBytes.Length < MinBytesPerWriteEvent)
             {
-                trailingBytes = trailingBytes.Concat(bytes.Skip(offset).Take(count)).ToArray();
+                this.trailingBytes = this.trailingBytes.Concat(bytes.Skip(offset).Take(count)).ToArray();
                 return;
             }
 
-            // if we have bytes left over from the previous write, prepend them
-            if (trailingByteLength > 0)
-            {
-                // convert the trailing bytes plus the first 3-trailingByteLength bytes of the new byte[]
-                prefixByteString = trailingBytes.Concat(bytes.Skip(offset).Take(numberOfBytesToPrefix)).ToArray();
-            }
+            byte[] byteArray;
 
-            // compute if there will be trailing bytes from this write
-            int remainingBytes = (count - numberOfBytesToPrefix) % 3;
-            trailingBytes = bytes.Skip(offset + count - remainingBytes).Take(remainingBytes).ToArray();
+            PrepareByteArray(bytes, offset, count, out byteArray);
 
-            JsonValueUtils.WriteBinaryString(this.Writer, prefixByteString.Concat(bytes.Skip(offset + numberOfBytesToPrefix).Take(count - numberOfBytesToPrefix - remainingBytes)).ToArray(), ref streamingBuffer, this.bufferPool);
+            JsonValueUtils.WriteBinaryString(this.Writer, byteArray, wrappedBuffer, this.bufferPool);
         }
 
-    
         /// <inheritdoc />
-        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        public override async Task WriteAsync(byte[] bytes, int offset, int count, CancellationToken cancellationToken)
         {
-            return TaskUtils.GetTaskForSynchronousOperation(() =>
-                this.Write(buffer, offset, count));
+            Debug.Assert(this.wrappedBuffer != null, "this.wrappedBuffer != null");
+
+            // if we have less than 3 bytes, store the bytes and continue
+            if (count + trailingBytes.Length < MinBytesPerWriteEvent)
+            {
+                this.trailingBytes = this.trailingBytes.Concat(bytes.Skip(offset).Take(count)).ToArray();

```

Comment:
Maybe this is not the scope of this PR, but just curious whether reallocating the `trailingBytes` all the time doesn't put pressure on the GC? Can a single array with an index to keep track of where to add new bytes? Or an array pool?

Reply:
And the impact of the LINQ calls?

Reply:
@habbes I stumbled upon this method and my thoughts were the same as yours. I did a refactor to make some of the logic shareable between `Write` and `WriteAsync` method.
What I believe is needed is a rewrite taking into consideration the suggestions you have provided. I even left a comment to that effect in the `PrepareByteArray` method where I moved the shareable code.
I'll have to make a judgement whether to create a separate issue for that or make that change in the PR. That said, I don't think this method is in a hot path based on the typical scenarios that this class is applied.